lib/toccatore/usage_update.rb

```diff
@@ -0,0 +1,171 @@
+
+      # return number of works queued
+      total
+    end
+
+    def process_data(options = {})
+      message = get_message
+      data = get_data(message)
+      data = parse_data(data, options)
+
+      return [OpenStruct.new(body: { "data" => [] })] if data.empty?
+
+      push_data(data, options)
+      delete_message message
+    end
+
+    def get_data reponse 
+      return OpenStruct.new(body: { "errors" => "Queue is empty" }) if reponse.messages.empty?
+
+      body = JSON.parse(reponse.messages[0].body)
+      Maremma.get(body["report_id"])
+    end
+
+
+    # method returns number of errors
+    def push_data(items, options={})
+      if items.empty?
+        puts "No works found in the Queue."
+        0
+      elsif options[:access_token].blank?
+        puts "An error occured: Access token missing."
+        options[:total]
+      else
+        error_total = 0
+        Array(items).each do |item|
+          error_total += push_item(item, options)
+        end
+        error_total
+      end
+    end
+
+    def metrics_url
+      "https://metrics.test.datacite.org/reports/"
+    end
+
+    def source_id
+      "usage_update"
+    end
+
+    def query
+      "relatedIdentifier:DOI\\:*"
+    end
+
+    def format_event type, data, options
+      { "id" => SecureRandom.uuid,
+        "message-action" => "add",
+        "subj-id" => data[:report_id],
+        "subj"=> {
+          "pid"=> data[:report_id],
+          "issued"=> data[:created]
+        },
+        "total"=> data[:count],
+        "obj-id" => data[:pid],
+        "relation-type-id" => type,
+        "source-id" => "datacite",
+        "source-token" => options[:source_token],
+        "occurred-at" => data[:created_at],
+        "license" => LICENSE 
+      }
+    end
+
+
+    def parse_data(result, options={})
+      return result.body.fetch("errors") if result.body.fetch("errors", nil).present?
+
+      items = result.body.dig("data","report","report-datasets")
+      header = result.body.dig("data","report","report-header")
+      report_id = metrics_url + result.body.dig("data","report","id")
+
+      created = header.fetch("created")
+      Array.wrap(items).reduce([]) do |x, item|
+        data = {}
+        data[:doi] = item.dig("dataset-id").first.dig("value")
+        data[:pid] = normalize_doi(data[:doi])
+        data[:created] = created
+        data[:report_id] = report_id
+        data[:created_at] = created
+
+        instances = item.dig("performance").first.dig("instance")
+
+        return x += [OpenStruct.new(body: { "errors" => "There are too many instances. There can only be 4" })] if instances.size > 8
+     
+        x += Array.wrap(instances).reduce([]) do |ssum, instance|
+          data[:count] = instance.dig("count")
+          event_type = "#{instance.dig("metric-type")}-#{instance.dig("access-method")}"
+          ssum << format_event(event_type, data, options)
```

Comment:
You probably need a new line with `ssum`.

Reply:
fixed 9c4c335